### PR TITLE
[rarun2] Added pseudo tty for listen and connect

### LINKED ISSUE #6303: rarun2 is buffering the ouput till the program ends
I am trying to redirect the io of a program to my script. So, I am using rarun2 with the listening option. But the output of my program is not sent to the script until the program has exited.

Filtering out the unnecessary details, I could distill the reproduction to this
```bash
#!/usr/bin/env bash

set -e

cat > io.c <<-C
#include <stdio.h>
int main() {
	int i;	printf("Started the binary\n");
	scanf("%d", &i);
	printf("Entered num %d\n", i);
	scanf("%d", &i);
	printf("Entered num %d\n", i);
}
C

gcc -o io io.c
rarun2 program=io listen=4000
```

In first terminal:

    $ ./issue.sh

In another terminal

    $ nc localhost 4000
    4
    5
    Started the binary
    Entered num 4
    Entered num 5

_4 and 5 are the inputs_

I think the expected output should be

    $ nc localhost 4000
    Started the binary
    4
    Entered num 4
    5
    Entered num 5

I am trying out rarun2 for the first time so I don't know if my use case is wrong. The intended use-case is to use it with r2 for debugging.

### PR DESCRIPTION
**Pull request by asutoshpalai**
This fixes output buffering for interactive programs when connected over socket.

Fixes #6303

### CODE REVIEW

File: libr/socket/run.c [hunk 1]
```diff
@@ -475,6 +475,100 @@ R_API const char *r_run_help() {
...
+
+	if (ioctl (in_fd, FIONREAD, &size) == -1) {
+		perror ("ioctl");
+		return -1;
+	}
+
+	if (size == 0) // child process exited or socket is closed
+		return -1;
+
+	*buff = realloc (*buff, size);
```
Comment (radare) [inline libr/socket/run.c]: dont realloc over the same buffer or memleaks can happen. use a temporary variable and check for null before going further

File: libr/socket/run.c [hunk 2]
```diff
@@ -475,6 +475,100 @@ R_API const char *r_run_help() {
...
+		FD_SET (sockfd, &readfds);
+
+		int activity = select (max_fd + 1, &readfds, NULL, NULL, NULL);
+
+		if (activity < 0) {
+			perror ("select error");
+			break;
+		}
+
+		if(FD_ISSET (fdm, &readfds)) {
```
Comment (radare) [inline libr/socket/run.c]: missing space

File: libr/socket/run.c [hunk 3]
```diff
@@ -475,6 +475,100 @@ R_API const char *r_run_help() {
...
+
+	if (write (out_fd, *buff, size) != size) {
+		perror ("write");
+		return -1;
+	}
+	return 0;
+}
+
+static int redirect_socket_to_pty (RSocket *sock) {
+#if __UNIX__ && !__ANDROID__ && LIBC_HAVE_FORK
```
Comment (radare) [inline libr/socket/run.c]: not sure how many other unixes restrict access to PTY. better use #if HAVE_PTY and define this variable at the begining of the file

File: libr/socket/run.c [hunk 4]
```diff
@@ -475,6 +475,100 @@ R_API const char *r_run_help() {
...
+	read (in_fd, *buff, size);
+
+	if (write (out_fd, *buff, size) != size) {
+		perror ("write");
+		return -1;
+	}
+	return 0;
+}
+
+static int redirect_socket_to_pty (RSocket *sock) {
```
Comment (radare) [inline libr/socket/run.c]: no space before ( in function definitions

**Comment by radare** [comment]
i think this feature should be enabled with pty=true in the rarun2 profile. not doing this by default (or at least allowing the user to choose if he wants to allocate a terminal to run the target)

**Comment by asutoshpalai** [comment]
Do I need to add documentation for `pty=true` anywhere other than the one shown with `rarun2 -h`

**Comment by radare** [comment]
<3

**Pull request merged**